src/pubsub/ua_pubsub_reader.c

```diff
@@ -1975,8 +1975,11 @@ receiveBufferedNetworkMessage(UA_Server *server, UA_ReaderGroup *readerGroup,
     while(buffer.length > currentPosition) {
         rv = decodeAndProcessNetworkMessageFun(
             server, readerGroup, connection, previousPosition, &buffer, &currentPosition);
-        UA_CHECK_STATUS_WARN(rv, return rv, &server->config.logger, UA_LOGCATEGORY_SERVER,
-                             "SubscribeCallback(): receive message failed");
+        if ((rv != UA_STATUSCODE_GOOD) && (rv != UA_STATUSCODE_BADNOTFOUND)) {
```

Comment:
1. We need to intercept the case when `decodeAndProcessNetworkMessageFun` did not progress in the buffer and no error was thrown, otherwise we might end up in an infinite loop (just add a first interception of the buffer progress with according logging and breaking the loop)
2. Would be great to create a separate if-clause for handling the `UA_STATUSCODE_BADNOTFOUND` specifically (including a specified logging message)
3. You can use UA_CHECK_STATUS_WARN for any other bad value that was encountered after the `BAD_NOT_FOUND` case

Reply:
Many thanks for your review!

I am not sure if I have understood all of your points:

1. I would have expected that the `decodeAndProcessNetworkMessageFun` would return an error if currentPosition is still 0. 
Shall I add an additional check, after checking the statusCode for BADNOTFOUND and other errors?
`if (previousPosition == currentPosition)`
But this hasn't been checked before I think -> but as you suggest!

2. I have added the if-clause but I print the log at `decodeAndProcessNetworkMessage` as `decodeAndProcessNetworkMessageRT` does. Is this ok?

Log looks like this:
```
[1601-01-01 00:02:48.545 (UTC+0000)] debug/server       PubSub subscribe callback
[1601-01-01 00:02:48.545 (UTC+0000)] info/server        Dataset reader not found. Check PublisherID, WriterGroupID and DatasetWriterID
[1601-01-01 00:02:48.545 (UTC+0000)] info/server        PubSub receive. Unknown message received. Will not be processed.

```

And I haven't used log category warning but info instead, because in the multicast scenario I think it's ok to receive these messages.

If you like and you think it's easier we could also talk about it

Reply:
I have rebased on current master and now it doesn't work anymore.

I think it has something to do with this commit: https://github.com/open62541/open62541/commit/b862f28e4bac4403c39c9cb5d9b436643cde3183

Reply:
because in the past the subscribeCallback() called UA_Server_processNetworkMessage -> which checked getReaderFromIdentifier -> which checked **all DataSetReaders of the ReaderGroup**

But now: decodeAndProcessNetworkMessage directly calls UA_ReaderGroup_processNetworkMessage, and this only checks the current readergroup
and then the full network message has been processed and the next readergroup of the same connection does not receive() anything anymore ...
maybe we just have to add 
    `LIST_FOREACH(readerGroup, &pConnection->readerGroups, listEntry) {`
    
    
    
In general I am not sure if it's a good idea to provide a subscribe callback for every ReaderGroup.
Because if there are multiple readergroups then (at least in the past) the first ReaderGroup received() everything and forwarded the network messages to the right readergroups and dataset readers within it's connection
the other ReaderGroups which were also scheduled had nothing to do in this case -> which causes unnecessary task switches in a realtime environment

I thought it would be simpler if there was one global reader task, which receives all messages for the connection and then forwards it to the correct readersgroups/datasetreaders

@JanSurft What do you think ?

Reply:
I don't understand, somehow it looks like Julius commit would have already solved all these issues

https://github.com/open62541/open62541/commit/b862f28e4bac4403c39c9cb5d9b436643cde3183

but it doesn't look like the current master branch

Reply:
Hmm, the commit removed UA_PubSubConnection_processNetworkMessage  and changed it to UA_ReaderGroup_processNetworkMessage, but I think the connection function is needed, even though we already know the readerGroup

Reply:
Just my thoughts (not sure if I have understood the current implementation correctly)

There's one socket for every connection -> that's ok

BUT: if there are multiple ReaderGroups, then every ReaderGroup has a SubscribeCallback
The ReaderGroup which is scheduled first calls connection->channel->receive() and reads the received data from the socket. -> With it's configured timeout. 

If all ReaderGroups have the same subscribe/corresponding publishing interval then the first 
ReaderGroup receives all network messages, even though they are meant for other ReaderGroups as well.
And then it does not forward the messages to the other ReaderGroups anymore.

And even if we say that it's unlikely that all ReaderGroups of a connection have the same subscribing interval, because it does not make a lot of sense to configure multiple WriterGroups with the same publishing interval, it could still be a problem if the interval difference is small enough and the network messages of other readergroups have already been received ...

Reply:
UA_ReaderGroup_subscribeCallback
receiveBufferedNetworkMessage
calls receive() -> reads everything that's available at the moment on the socket -> maybe also data for other ReaderGroups
decodeAndProcessNetworkMessage

- thinks that it only has network messages for this ReaderGroup
- param connection would be given to check all ReaderGroups, but there are problems with decodeNetworkMessage because verifyAndDecryptNetworkMessage also uses a specific ReaderGroup, UA_NetworkMessage_decodeHeaders would probably work if we call it for multiple ReaderGroups ...